### PR TITLE
KEP-1618 - don't add requests to newview if we don't have them

### PR DESCRIPTION
**Pull request by paularchard**
Fix for testnet-2 crash July 18 2019

### CODE REVIEW

File: pbft/pbft.cpp [hunk 1]
```diff
@@ -1307,7 +1307,7 @@ pbft::is_valid_newview_message(const pbft_msg& theirs, const bzn_envelope& origi
         viewchange_envelopes_from_senders[viewchange_env.sender()] = viewchange_env;
     }
 
-    pbft_msg ours = this->build_newview(theirs.view(), viewchange_envelopes_from_senders);
+    pbft_msg ours = this->build_newview(theirs.view(), viewchange_envelopes_from_senders, false);
     if (ours.pre_prepare_messages_size() != theirs.pre_prepare_messages_size())
```
Comment (rnistuk) [inline pbft/pbft.cpp]: Cool, while checking for new view message validity we do not need to worry about adding requests we don't have.

File: pbft/pbft.cpp [hunk 2]
```diff
@@ -1470,7 +1471,7 @@ pbft::build_newview(uint64_t new_view, const std::map<uuid_t,bzn_envelope>& view
             }
 
             auto env = this->wrap_message(pre_prepare);
-            if (pre_prepare.request_type() == pbft_request_type::PBFT_REQUEST_PAYLOAD)
+            if (pre_prepare.request_type() == pbft_request_type::PBFT_REQUEST_PAYLOAD && attach_reqs)
```
Comment (rnistuk) [inline pbft/pbft.cpp]: OK, here's where we attach the requests...

File: pbft/pbft.cpp [hunk 3]
```diff
@@ -1480,7 +1481,15 @@ pbft::build_newview(uint64_t new_view, const std::map<uuid_t,bzn_envelope>& view
                 }
                 auto op = this->operation_manager->find_or_construct(old_view, pre_prepare.sequence()
                     , pre_prepare.request_hash(), config->get_peers());
```
Comment (rnistuk) [inline pbft/pbft.cpp]: Cool, checking that the operation has a request before attempting to access it.

File: pbft/pbft.hpp [hunk 4]
```diff
@@ -229,7 +229,7 @@ namespace bzn
         void initiate_viewchange();
         std::shared_ptr<bzn_envelope> make_viewchange(uint64_t new_view, uint64_t n, const std::unordered_map<bzn::uuid_t, std::string>& stable_checkpoint_proof, const std::map<uint64_t, std::shared_ptr<bzn::pbft_operation>>& prepared_operations);
         pbft_msg make_newview(uint64_t new_view_index,  const std::map<uuid_t,bzn_envelope>& viewchange_envelopes_from_senders, const std::map<uint64_t, bzn_envelope>& pre_prepare_messages) const;
-        pbft_msg build_newview(uint64_t new_view, const std::map<uuid_t,bzn_envelope>& viewchange_envelopes_from_senders) const;
+        pbft_msg build_newview(uint64_t new_view, const std::map<uuid_t,bzn_envelope>& viewchange_envelopes_from_senders, bool attach_reqs = false) const;
```
Comment (rnistuk) [inline pbft/pbft.hpp]: Based on the usage pattern of build_newview, that is the only time attach_reqs is actually set, it is set to false, is false the correct default here?

File: pbft/pbft.cpp (hunk 1 above)
Comment (paularchard) [reply pbft/pbft.cpp]: No the default should be true. Changing it.

File: pbft/pbft.hpp (hunk 4 above)
Comment (rnistuk) [reply pbft/pbft.hpp]: Cool! Fixed it. Thanks.

**Pull request merged**